### ISSUE TITLE
Using caching to speed up the builds.

### ISSUE THREAD

**Issue by coryan** [GoogleCloudPlatform/google-cloud-cpp#118]
This should probably happen after #49.  We should cache the build output for faster builds, specially because so much of the time in the build is spent on dependencies (`third_party/grpc` in particular).

The caching should take into account the compiler flags, as described in the Travis [documentation](https://docs.travis-ci.com/user/caching/#Caches-and-build-matrices)

**Comment by coryan** [comment]
No success so far with speeding up the builds using caching.  @mbrukman this is not urgent, but if you can take a look and point out any mistakes or suggestions that would be good.  I have ran out of ideas.

The changes can be examined here:

https://github.com/GoogleCloudPlatform/google-cloud-cpp/compare/master...coryan:cache-build

the effect on the build is none?  Compare the build on master:

https://travis-ci.org/coryan/google-cloud-cpp/builds/322440585

against the build with caching enabled:

https://travis-ci.org/coryan/google-cloud-cpp/builds/322641137

I do not see a substantial difference, and the debug output shows that the caching build is downloading the cache and putting it in the right place.  Possibly make/cmake are too clever and they rebuild anyway because the Makefiles or some of the headers changed?

In any case, out of ideas, welcome suggestions, will probably close as 'wontfix' if we cannot figure out a simple solution here.

**Comment by coryan** [comment]
Note that #603 fixed this for the Bazel-based builds, and #611 fixed it for most builds.  Also #614 will fix this for the builds that install dependencies.

That leaves only two cases:

- [ ] The macOS builds are not using ccache.
- [ ] The code coverage builds are not using ccache because everything breaks when I tried to use it :disappointed: 

Will leave this open for the moment, but we should probably close it and open new specific bugs.

**Comment by coryan** [comment]
I do not expect I will be making any progress on this for the time being. Only the Coverage builds do not use caching right now.

**Comment by coryan** [comment]
Fixed with #1292.
